cpu/samd21/cpu.c

```diff
@@ -31,12 +40,27 @@ static void clk_init(void)
     PM->APBAMASK.reg = (PM_APBAMASK_PM | PM_APBAMASK_SYSCTRL |
                         PM_APBAMASK_GCLK);
 
-    /* adjust NVM wait states, see table 42.30 (p. 1070) in the datasheet */
-#if (CLOCK_CORECLOCK > 24000000)
+    /* adjust NVM wait states, see SAMD21 datasheet
+       Rev A (2017) table 37-40 , page 816 */
     PM->APBBMASK.reg |= PM_APBBMASK_NVMCTRL;
+#if (VDD_MILLIVOLTS > 2700)
+#if (CLOCK_CORECLOCK > CLOCK_24MHZ)
     NVMCTRL->CTRLB.reg |= NVMCTRL_CTRLB_RWS(1);
-    PM->APBBMASK.reg &= ~PM_APBBMASK_NVMCTRL;
+#else
+    NVMCTRL->CTRLB.reg |= NVMCTRL_CTRLB_RWS(0);
 #endif
+#else /* VDD_MILLIVOLTS <= 2700 */
+#if (CLOCK_CORECLOCK > CLOCK_42MHZ)
+    NVMCTRL->CTRLB.reg |= NVMCTRL_CTRLB_RWS(3);
+#elif (CLOCK_CORECLOCK > CLOCK_28MHZ)
+    NVMCTRL->CTRLB.reg |= NVMCTRL_CTRLB_RWS(2);
+#elif (CLOCK_CORECLOCK > CLOCK_14MHZ)
+    NVMCTRL->CTRLB.reg |= NVMCTRL_CTRLB_RWS(1);
+#else
+    NVMCTRL->CTRLB.reg |= NVMCTRL_CTRLB_RWS(0);
+#endif
```

Comment:
Seems awfully complicated to me. How about you simply add this to the top of the file:
```c
#ifndef VDD
/**
 * @brief   Set default system voltage level in mV (used for determine flash wait states)
 * 
 * @note    Override this value in your boards periph_conf.h file if a different system
 *               voltage is used.
 */
#define VDD                 (3300U)
#endif

/* determine the needed flash wait states based on the system voltage (Vdd) */
#if (VDD > 2700)
#define WAITSTATES          ((CLOCK_CORECLOCK - 1) / 24000000)
#else
#define WAITSTATES          ((CLOCK_CORECLOCK - 1) / 14000000)
#endif
```
And then change the actual editing of the CTRLB register to:
```c
NVMCTRL->CTRLB.reg |= NVMCTRL_CTRLB_RWS(WAITSTATES);
```

This should fully suffice, leave out those `CLOCK_xxMHZ` defines. IHMO we could also drop the `VDD` defines from the boards periph_conf.h files, as they fallback